packages/typegpu/src/core/root/init.ts

```diff
@@ -289,14 +289,54 @@ class TgpuRootImpl extends WithBindingImpl
     super(() => this, []);
 
     this['~unstable'] = this;
+
+    let commandEncoder: GPUCommandEncoder | undefined;
+    this[$internal] = {
+      batchState: {
+        ongoingBatch: false,
+        performanceCallbacks: [],
+      },
+
+      get commandEncoder() {
+        if (!commandEncoder) {
+          commandEncoder = device.createCommandEncoder();
+        }
+
+        return commandEncoder;
+      },
+
+      flush() {
+        if (!commandEncoder) {
+          return;
+        }
+
+        device.queue.submit([commandEncoder.finish()]);
+        commandEncoder = undefined;
+      },
+    };
   }
 
-  get commandEncoder() {
-    if (!this._commandEncoder) {
-      this._commandEncoder = this.device.createCommandEncoder();
+  batch(callback: () => void) {
+    const nestedBatch = this[$internal].batchState.ongoingBatch;
+    if (nestedBatch) {
+      throw new Error('Nested batch is not allowed');
     }
 
-    return this._commandEncoder;
+    this[$internal].batchState.ongoingBatch = true;
+
+    try {
+      callback();
+    } finally {
+      this[$internal].flush();
+      for (
+        const performanceCallback of this[$internal].batchState
+          .performanceCallbacks
+      ) {
+        performanceCallback();
+      }
+      this[$internal].batchState.ongoingBatch = false;
+      this[$internal].batchState.performanceCallbacks = [];
+    }
```

Comment:
Stylistic preference - a little less member accesses
```suggestion
    const { batchState } = this[$internal];

    if (batchState.ongoingBatch) {
      throw new Error('Nested batch is not allowed');
    }

    batchState.ongoingBatch = true;

    try {
      callback();
    } finally {
      this[$internal].flush();

      for (const performanceCallback of batchState.performanceCallbacks) {
        performanceCallback();
      }

      batchState.ongoingBatch = false;
      batchState.performanceCallbacks = [];
    }
```